### PR TITLE
Add support of ARC MWDT libc to Zephyr

### PR DESCRIPTION
**Pull request by evgeniy-paltsev**
ARC MWDT toolchain doesn't provide NEWLIB libc. Let's add support of of ARC MWDT libc to Zephyr.

This is prerequisite of adding ARC MWDT C++ library support to Zephyr.

### CODE REVIEW

**Comment by abrodkin** [comment]
Needs to be re-based on top of https://github.com/zephyrproject-rtos/zephyr/pull/36744 once it is merged.

**Review by abrodkin — APPROVED** [review]

File: lib/libc/arcmwdt/arcmwdt-string.c
Comment (ruuddw) [inline lib/libc/arcmwdt/arcmwdt-string.c:20]: No strnlen in MW libc? maybe file a enhancement request for that...
I do see strnlen_s is part of MW, use that instead?

File: lib/libc/arcmwdt/include/fcntl.h
Comment (ruuddw) [inline lib/libc/arcmwdt/include/fcntl.h:10]: Aren't these define in MW sys/fctl?
Can you simply define them here, without ensuring that the MW implementation isn't using other constants?

File: lib/libc/arcmwdt/arcmwdt-string.c
Comment (evgeniy-paltsev) [reply lib/libc/arcmwdt/arcmwdt-string.c:13]: Fixed. Now we simply call strnlen_s in strnlen.

File: lib/libc/arcmwdt/include/fcntl.h
Comment (evgeniy-paltsev) [reply lib/libc/arcmwdt/include/fcntl.h]: Fixed. Now we only define O_NONBLOCK which is missing in MWDT fcntl.h header. We also check that our O_NONBLOCK value doesn't conflict with any other file IO defines which came from MWDT fcntl.h header.

Comment (ruuddw) [reply lib/libc/arcmwdt/include/fcntl.h]: Nice solution, thanks.

**Review by ruuddw — APPROVED** [review]
Ready for merge imho.

**Comment by evgeniy-paltsev** [comment]
The warnings in compliance check (`WARNING:NEW_TYPEDEFS: do not add new typedefs`) are related to adding types missing in MWDT toolchain, so they should be ignored.

**Review by dcpleung — APPROVED** [review]

**Comment by evgeniy-paltsev** [comment]
This PR has enough approvals and it was kept open enough time so everyone could share their concerns (if any). I guess we can safely merge it now.

File: lib/libc/arcmwdt/libc-hooks.c
Comment (stephanosio) [inline lib/libc/arcmwdt/libc-hooks.c:31]: Here you are implicitly implementing the "minimal libc" hook interface functions:
https://github.com/zephyrproject-rtos/zephyr/blob/775cf55d9268201be6f42523c80ffad808fd3209/include/sys/libc-hooks.h#L19-L37

You should add `#elif defined(CONFIG_ARCMWDT_LIBC)` there and add only the required hook functions for the ARC MWDT libc.

**Review by stephanosio — CHANGES_REQUESTED** [review]
The `libc-hooks.c` implementation does not look sane to me.

I assume `fputs` and others are already implemented by the ARC MWDT libc, but they are declared weak and therefore overridable.

In that case, only the most fundamental (interface) functions should be overridden by the `libc-hooks.c` (e.g. `write()` and `read()` as with the newlib implementation). I highly doubt the ARC MWDT libc implementation just happens to have the same interface definition as what we currently have in `stdout_console.c` of the minimal libc.

File: lib/libc/arcmwdt/libc-hooks.c
Comment (evgeniy-paltsev) [reply lib/libc/arcmwdt/libc-hooks.c]: Yep, you are correct. I've dropped it in a new implementation.

File: lib/libc/arcmwdt/libc-hooks.c
Comment (evgeniy-paltsev) [reply lib/libc/arcmwdt/libc-hooks.c]: Fixed in new implementation, thanks!

File: include/toolchain/mwdt.h
Comment (evgeniy-paltsev) [reply include/toolchain/mwdt.h]: Removed.

File: lib/libc/arcmwdt/libc-hooks.c
Comment (evgeniy-paltsev) [reply lib/libc/arcmwdt/libc-hooks.c]: Fixed.

**Review by stephanosio — CHANGES_REQUESTED** [review]
Looks ok overall. Some minor fixes.

File: lib/libc/arcmwdt/libc-hooks.c
Comment (stephanosio) [inline lib/libc/arcmwdt/libc-hooks.c:48]: Cast should not be necessary.
```suggestion
	return z_impl_zephyr_write_stdout(buf, nbytes);
```

Comment (evgeniy-paltsev) [reply lib/libc/arcmwdt/libc-hooks.c]: Thanks for pointting, fixed!

File: lib/libc/arcmwdt/libc-hooks.c
Comment (stephanosio) [inline lib/libc/arcmwdt/libc-hooks.c:67]:
```
-:308: ERROR:OPEN_BRACE: open brace '{' following function definitions go on the next line
#308: FILE: lib/libc/arcmwdt/libc-hooks.c:66:
+int _isatty(int file) {
```

```suggestion
int _isatty(int file)
{
	if (file == STDIN_FILENO || file == STDOUT_FILENO || file == STDERR_FILENO) {
```

**Review by stephanosio — APPROVED** [review]

**Review by ruuddw — APPROVED** [review]
re-approved latest version

**Comment by cfriedt** [comment]
Thanks @evgeniy-paltsev - will merge as soon as it's green. And thanks to everyone for the reviews ;-)

**Comment by evgeniy-paltsev** [comment]
@cfriedt ok, now there are only a few warnings about adding typedefs missing in MWDT toolchain, so they should be ignored.

Thanks for pointing to the issue with too long commit line!

**Pull request merged**
